### ISSUE TITLE
[IDE0069] Disposable fields when passed from the outside should not be disposed

### ISSUE THREAD

**Issue by SergeyTeplyakov** [dotnet/roslyn#38900]
**Version Used**: 
Visual Studio 16.3.0

**Steps to Reproduce**:

According to [CA2213](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2213-disposable-fields-should-be-disposed?view=vs-2019) only disposable fields that were created locally should be disposed inside the `Dispose` method. But **If the object is created or assigned outside of type T, the rule does not fire. This reduces noise for cases where the containing type doesn't own the responsibility for disposing of the object.**.

It means that the following code should not trigger IDE0069:

```csharp
internal class FooBar : IDisposable
{
    private readonly Stream _d;

    public FooBar(Stream d)
    {
        _d = d;
    }

    public void Dispose()
    {
        throw new NotImplementedException();
    }

    public override string ToString()
    {
        return _d.ToString();
    }
}
```

But it does.

**Expected Behavior**:

`FooBar` should not have any IDE0069 diagnostics.

**Actual Behavior**:

The IDE0069 diagnostic is emitted for disposable fields passed via constructor

**Comment by sharwell** [comment]
Disposable arguments to a constructor are often treated as transferring ownership to the newly-constructed type. If the non-copyable structs proposal gets implemented, it will be possible to make these cases explicit in an API.

**Comment by SergeyTeplyakov** [comment]
> Disposable arguments to a constructor are often treated as transferring ownership to the newly-constructed type. If the non-copyable structs proposal gets implemented, it will be possible to make these cases explicit in an API.

I don't think this is true. I would say that this is a rare case from my experience (and this is the reason why the exceptions list in the article that I've mentioned is so short).

I'm happy to do the research to prove my point, but my bet is that ownership transfer would be in minority;). And this is, I think, quite reasonable: there is a subtle difference between composition and aggregation. Aggregation means that one instance depends on each other in one way or another. Usually, this relationship is expressed by passing a reference into a method/constructor. The composition is a stricter form of aggregation when one object controls the lifetime of another.

In my opinion, the rule should be triggered only in the case of composition, when "the whole" actually control the lifetime of "the piece" by constructing the field by itself.

Moreover, I truly believe that the diagnostic in the current form is potentially harmful, because people can start disposing instances that they don't control without thinking about the ownership too much. 

And, btw, the current implementation clearly violates the rule stated in FDGL:

```
Disposable fields should be disposed This usage rule reports a warning when a type creates and owns disposable instances but does not call Dispose on these instances in its Dispose method. Types are considered as owning a disposable instance when they assign an instance they have explicitly created to one of their instance fields.
```

P.S. I know that CA* rules are effectively enforce FDGL, but my point is - if the guideline has changed, I would love to understand why it was changed and based on what data. Because my experience matches the guidelines 100%:)

**Comment by sharwell** [comment]
Examples of transfer would include:

* `StreamReader`
* `ReferenceCountedDisposable<T>`
* `System.Net.Http.StreamContent`

My main objection to the rules today is there is no guidance available which a user can follow to end up with correct code and no false positives or false negatives from the analyzer.

**Comment by SergeyTeplyakov** [comment]
I'm not saying there are no exceptions to my point. What I'm saying is that the transfer is an exception. And the IDE rules should be built around the common case.

> My main objection to the rules today is there is no guidance available which a user can follow to end up with correct code and no false positives or false negatives from the analyzer.

I disagree with this: the rule is - dispose the stuff that you create. And the diagnostic should detect this case and should not cause false negatives in the case when the resource is most likely not owned by the instance.

From my perspective, the analyzer should be trustworthy and more conservative if needed otherwise it'll be disabled.

I don't have any other options right now - I have to globally disable the rule because it yield too much noise. But because the analyzer is so aggressive I won't see the legit cases when the resources that owned exclusively by some type are not disposed.

**Comment by CyrusNajmabadi** [comment]
> What I'm saying is that the transfer is an exception. 

I'm not certain that's true.  'Stream' is the most common case of IDisposable I ever use.  It's enormously widespread, and has followed this convention since the beginning of .net.  It's what i think of as the way to do disposable.  

> From my perspective, the analyzer should be trustworthy and more conservative if needed otherwise it'll be disabled.

This is always a personal preference.  Some will say: it didn't find this case over here, so i don't trust it.  It should be less conservative to find all the cases, and i can disable it in the cases it misses. 

It's hard to meet the needs of everyone in this space.

**Comment by SergeyTeplyakov** [comment]
> I'm not certain that's true. 'Stream' is the most common case of IDisposable I ever use. It's enormously widespread, and has followed this convention since the beginning of .net. It's what i think of as the way to do disposable.

Hm.. I don't think that `Stream` actually transfers the onwership. `StreamReader` definitely does, but classes like `FileStream` usually (with some exceptions) own the underlying resources but not take them via a constructor.

And this use case should not be analyzed based on whether the type that follows or not some specific pattern is widely used or not. It doesn't matter. We should analyze the number of types that transfer the ownership in the wild vs. the number of types that just reference the resource passed via the constructor and the outer code owns the ownership.

> This is always a personal preference. Some will say: it didn't find this case over here, so i don't trust it. It should be less conservative to find all the cases, and i can disable it in the cases it misses.

I have FDGL on my side:) and if something has changed and the guideline is no longer valid, then I would like to see some evidence :). Data-driven evidence. I believe that FDGL were built based on excessive analysis of existing codebases and we should have reasonable counter evidence to stop following the rules (I definitely see enough evidence to not follow the dispose pattern itself, but this is a different story). 

And I don't think that there are any expectations from the analyzers to find "all" (or even majority) of error cases, but I definitely expect from the analyzer to show as few false positives as possible.

**Comment by mavasani** [comment]
Let me clarify a few things here:

1. Both CA2213 and IDE0069 have been designed to follow identical semantics here. IDE0069 may diverge where reasonable, but the issue in this thread does not belong to that category.
2. In general the statement made by @SergeyTeplyakov in the issue description is true and matches the documentation of CA2213. However, the documentation also clearly states the special cases where the rule diverges from these semantics, `Stream` being one of those types: https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2213-disposable-fields-should-be-disposed?view=vs-2019#special-cases. So, if you modify your example to have a field of any IDisposable type other then the 4 special ones noted in the documentation, neither CA2213 nor IDE0069 will fire.

> It's hard to meet the needs of everyone in this space.
I completely agree with this assessment. Dispose ownership transfer at call sites is a tricky subject, and  IMO the best path forward is to provide suitable options to allow end users to configure their preference. For example, for CA2000 (DisposeObjectsBeforeLosingScope), we introduced couple of options [here](https://github.com/dotnet/roslyn-analyzers/blob/master/docs/Analyzer%20Configuration.md#configure-dispose-ownership-transfer-for-arguments-passed-to-constructor-invocation) and [here](https://github.com/dotnet/roslyn-analyzers/blob/master/docs/Analyzer%20Configuration.md#configure-dispose-ownership-transfer-for-disposable-objects-passed-as-arguments-to-method-calls), to allow end user configuration. We should possibly do something similar for CA2213 and IDE0069 to allow end users to tweak the analysis to match their requirements.

**Comment by SergeyTeplyakov** [comment]
@mavasani thanks a lot for the explanation! Now the behavior makes sense.

**Comment by mavasani** [comment]
> My main objection to the rules today is there is no guidance available which a user can follow to end up with correct code and no false positives or false negatives from the analyzer.

I have disagreed with Sam for quite a while on this. "no false positives or false negatives from the analyzer" requirements cannot be met by any dataflow analysis based rule that needs complete interprodedural dataflow analysis for correctness, until it actually perform such as expensive analysis (very unlikely). For example, CA2213 is well documented along with special cases at https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2213-disposable-fields-should-be-disposed?view=vs-2019#rule-description and it is inevitable that the rule cannot match the above requirement because the semantics are very much dependent on how code flow occurs, which either needs complete interprocedural dataflow analysis OR needs users to annotate their code and/or use options to reduce their false positives. Going to pessimistic in analysis can lead to large number of false negatives to drastically reduce the rules usefulness. So generally, the analysis has to choose a balance. The rules even with some false positives are still valuable as they catch real performance and security issues, with the known overhead of navigating through some of these false positives.
